### ISSUE TITLE
OTOTOY

### ISSUE THREAD

**Issue by UltimateRiff** [kellnerd/harmony#34]
I honestly couldn't find any API for OTOTOY, but since it is a Japanese store, most of the help pages aren't in English, so there might be.

https://ototoy.jp

that said, perhaps it could be scraped for data, especially since it's one of the few stores I know that shows catalog numbers (for example, [here](https://ototoy.jp/_/default/p/637303)).

important note, OTOTOY does keep seperate pages for Lossless and High-Resolution releases, which would be the same MusicBrainz release (all other data being the same, of course)

**Comment by UltimateRiff** [comment]
also, I believe releases _might_ only be available in Japan, but I'm not 100% certain on that... I think they could be available in Europe too? again, not sure about the borders of availability (and if they change from release to release)

**Comment by julian45** [comment]
OTOTOY is restricted to the Japanese market only. Furthermore, they don't appear to have an API either.

That being said, they're a decent source of cover art; sometimes, they have even higher-resolution cover art than, e.g., iTunes. If there's any interest in pulling that in, I think it's programmatically possible to find the link to the full-resolution cover art; I believe one of the "Music Hoarders" userscripts that deals with cover art has OTOTOY as a provider. (I'd find the link but am away from my computer for the time being.)

**Comment by julian45** [comment]
TL;DR: found what I was talking about, but it relies on a centralized service that does not appear to provide source code. Might be worth that trouble to get cover art integration; might not. In any case, it appears possible to scrape the tracklist, and given a (scraped) link to `imgs.ototoy.jp`, you might be able to extract the cover art image yourself.

I was able to track down the userscript I mentioned before that uses a Music Hoarders project to get high-quality cover art from OTOTOY. The Git host for it is currently experiencing some issues, but I extracted a direct link to the script from my browser's copy:
https://notabug.org/SuperSaltyGamer/ame/raw/main/dist/musicbrainz.user.js
The repo itself would most likely be found at: https://notabug.org/SuperSaltyGamer/ame

IIRC the Ame project has a number of integrations with other sites that could be useful (or not) for this project, but this particular script, designed for use on MB's cover art addition page, basically just seeds some data about the release to [this closed-source project](https://covers.musichoarders.xyz/)'s interface, then takes the cover art the user selects (and the project feeds it) and drops it into MB's image upload form. The "integration" link near the top of that page details how to programmatically interact with the project. IMO, the project has some benefits, but the fact that it (at least by appearances) is closed-source would not make me comfortable basing an application function on it if I were in kellnerd's shoes.

That being said, for cover art scraping, it might be possible to benefit from [Image Max URL](https://github.com/qsniyg/maxurl) or the techniques it itself uses for OTOTOY. As for tracklist scraping, there might be some benefit in adopting techniques used in [this userscript](https://greasyfork.org/en/scripts/442490-mb-copy-tracklist-from-ototoy).

**Comment by julian45** [comment]
> I've also seen similar releases here in the US, so not _entirely_ Japan-only, only mostly

Right — quite often, there will be releases that are available worldwide on various other streaming & download sites, but OTOTOY's (almost) sole user market & user base is in Japan. (same with, e.g., mora and recochoku/レコチョク)

**Comment by UltimateRiff** [comment]
no, I mean streamable _on OTOTOY_, not on other streaming services

**Comment by Serial-ATA** [comment]
@kellnerd I can work on this

**Comment by kellnerd** [comment]
@Serial-ATA Is there any way to get structured data from OTOTOY?
I surely don't want to maintain a pure HTML DOM scraper. That is already an unreliable mess as a userscript implementation, which gets even worse with Harmony having to maintain backwards compatibility for old snapshots, otherwise permalinks are going to break.
